### PR TITLE
ROOT: use native CMakePackage functions in packages.py

### PR DESCRIPTION
**Pull request by sethrj**
This changes the ROOT package to use the newish `CMakePackage.define` and `CMakePackage.define_from_variant` instead of a custom `_process_opts` script. Since `define_from_variant` checks the existence of the variant name, it discovered that four configuration options (minimal, odbc, roofit, veccore) were implicitly being set to False. For the time being I've preserved the existing behavior by explicitly defining those to false, but I'd be happy to add the variants, change the defaults, or remove the definitions as needed. Thanks!

### CODE REVIEW

**Review by adamjstewart — APPROVED** [review]
@chissg @HadrienG2 @drbenmorgan @vvolkl can you test this?

**Comment by sethrj** [comment]
FWIW I did just install
```
root@6.20.08+aqua~davix~emacs~examples~fftw~fits~fortran+gdml+gminimal~graphviz~gsl~http~ipo~jemalloc~math~memstat+minuit~mlp~mysql~opengl~postgres~pythia6~pythia8+python~qt4~r~root7~rootfit+rpath~shadow~sqlite~ssl~table~tbb+threads~tmva+unuran~vc+vdt~vmc~x+xml~xrootd build_type=RelWithDebInfo cxxstd=14 patches=22af3471f3fd87c0fe8917bf9c811c6d806de6c8b9867d30a1e3d383a1b929d7 darwin-catalina-x86_64 / apple-clang@12.0.0
```
and it all looks good, but I do want to get input from the HEP maintainers about the potential missing variants.

File: var/spack/repos/builtin/packages/root/package.py
Comment (HadrienG2) [inline var/spack/repos/builtin/packages/root/package.py]:
```suggestion
            options.append(define('CMAKE_PROGRAM_PATH',
```

**Review by HadrienG2 — CHANGES_REQUESTED** [review]
I do not have the time to run a test build at the moment, but I have carefully read this PR and I believe it is indeed a faithful translation from the old homemade _process_opts method to the new CMakePackage define* methods. Except for the little CMAKE_PROGRAM_PATH typo mentioned as a code comment.

To answer your question about variants, I would personally add variants for `odbc`, `roofit` and `veccore`, but I don't think it is a good idea to add a variant for `minimal` because it would conflict with so many other variants.

I'll let @chissg and @drbenmorgan comment further as they see fit.

File: var/spack/repos/builtin/packages/root/package.py
Comment (sethrj) [reply var/spack/repos/builtin/packages/root/package.py]: Good catch, thanks!

File: var/spack/repos/builtin/packages/root/package.py
Comment (sethrj) [inline var/spack/repos/builtin/packages/root/package.py]: Actually @HadrienG2 this paragraph might not be necessary at all -- it's just adding the mysql-client client's install path to the CMake program search path, but since Spack sets CMAKE_PREFIX_PATH already based on the dependent specs, and that variable is also used to search for executables, I think setting CMAKE_PROGRAM_PATH is redundant.

**Comment by sethrj** [comment]
- There's a `README.md` sitting next to the ROOT packagefile, and it lists `odbc` under  "permanently removed variables", so I think we can leave that one set to False. 
- The `veccore` option I think is redundant with the `vc` -- at least, it looks like veccore installs vc (optional?).
- I peeked a bit into the "roofit" option -- it's a feature that looks like it has a number of dependencies on the selected package, so that may be risky to add as a variant unless someone's willing to test.

I'll wait for the other package authors for further instructions.  Thanks for the initial review!

File: var/spack/repos/builtin/packages/root/package.py
Comment (drbenmorgan) [reply var/spack/repos/builtin/packages/root/package.py]: Hi @sethrj, I think it is safe to remove this. ROOT supplies its own `FindMySQL.cmake`, and my reading of this section: 

https://github.com/root-project/root/blob/master/cmake/modules/FindMySQL.cmake#L18-L31 

suggests it'll respect `CMAKE_PREFIX_PATH`.

**Review by drbenmorgan — APPROVED** [review]
I also read this as a perfect translation of the current options/variants, so good to go from my perspective. I think adding variants can/should be dealt with in a separate PR.

**Comment by drbenmorgan** [comment]
Ahhh, sorry @sethrj, if you want to go ahead and remove the `mysql-client` related paragraph that's fine. I'll merge after that or let me know if you want to leave in place.

**Comment by drbenmorgan** [comment]
Thanks for the update @sethrj, good for merge!

**Pull request merged**
